go-present-mode.el

```diff
@@ -0,0 +1,1278 @@
+    (define-key map (kbd "C-c C-p C-l") #'go-present-insert-edit-link)
+    (define-key map (kbd "C-c C-p *") #'go-present-toggle-font-bold)
+    (define-key map (kbd "C-c C-p _") #'go-present-toggle-font-italic)
+    (define-key map (kbd "C-c C-p `") #'go-present-toggle-font-program)
+    (define-key map (kbd "C-c C-p C-v") #'go-present-show-code)
+    (define-key map (kbd "C-c C-p C-c") #'go-present-run-code-at-point)
+    (define-key map [remap save-buffer] #'go-present-save-buffer)
+    (define-key map (kbd "<M-return>") #'go-present-insert-slide)
+    map)
+  "Keymap for `go-present-mode'.")
+
+(defun go-present-insert-slide ()
+  "Insert a new slide."
+  (interactive)
+  (cond
+   ((or (bobp) (bolp))
+    (insert "* "))
+   ((eobp) (insert "\n* "))
+   ((go-present--at-code (point) (1+ (point)))
+    (error "Can't insert slide inside code section"))
+   ((looking-at-p "^\\*+\\s-+")
+    (insert "\n")
+    (forward-line -1)
+    (insert "* "))
+   ((looking-back "^\\*+\\s-+")
+    (end-of-line)
+    (insert "\n* "))
+   (t (insert "\n* ")
+      (end-of-line))))
+
+(defun go-present--url-near-point ()
+  "Return the URL at point.
+
+If none, return the first url after point, on the same line."
+  (car (or (get-text-property (point) 'go-present-link)
+           (get-text-property
+            (or (next-single-property-change
+                 (point) 'go-present-link nil (point-at-eol)) (point))
+            'go-present-link))))
+
+(defun go-present--find-code-other-window (location &optional offset)
+  "Open code location in a other window.
+
+LOCATION is a dotted pair of file name and Go present address
+string to seek to in the file.
+After moving the point to LOCATION, move it forward by OFFSET
+characters."
+  (let ((slide-buf (current-buffer)))
+    (find-file-other-window (car location))
+    (go-present--add-code-modification-hook (current-buffer) slide-buf))
+  (when (cdr location)
+    (goto-char (car (go-present--addr-str-to-region (cdr location))))
+    (forward-char (or offset 0))))
+
+(defconst go-present--image-regexp
+  "^\\.\\(?:image\\|background\\)\\s-+\\(.*?\\)\\(?:\\s-+\\([0-9_]+\\)\\s-+\\([0-9_]+\\)\\)?$"
+  "Regular expression matching .image and .background commands.
+
+Subexpressions are:
+  1. file name
+  2. optional height (number or '_')
+  3. optional widith (number or '_')")
+
+(defconst go-present--link-regexp
+  "^\\(\\.link\\|\\.iframe\\)\\s-+\\([^\t\v\n\f ]+\\)"
+  "Regular expression matching .link and .iframe commands.
+
+Subexpression 1 the command (.link or .iframe).
+Subexpression 2 is the file name.")
+
+(defconst go-present--html-regexp "\\.html\\s-+\\(.*\\)$"
+  "Regular expression matching .html command.
+
+Subexpression 1 is the file name.")
+
+(defconst go-present--code-start-regexp "^\\.\\(code\\|play\\)\\s-+"
+  "A ragular expression matching start of a code section.
+
+Subexpression 1 is the command (.code or .play).")
+
+(cl-defun go-present-open-at-point ()
```

Comment:
I think that it can, but that won't simplify the code.
Currently I can do a regexp match and then pick a submatch as the result (and give it to a function like browse-url).
The alternative is to call thing-at-point (after I define forward/backward functions), do a regexp match on the result, and then pick submatch.

I can update the code if you think that thing-at-point is a better approach (it might have benefits that I'm not aware of).